File: python/tvm/relax/distributed/distributed_tensor.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
# pylint: disable=redefined-builtin, invalid-name
"""Data structures for distributed tensor."""
import tvm
from typing import List, Union, Tuple
from tvm.ir.global_info import GlobalInfo
from tvm.runtime import ShapeTuple
from tvm.runtime.object import Object

from . import _ffi_api as ffi


@tvm._ffi.register_object("relax.distributed.DeviceMesh")
class DeviceMesh(GlobalInfo):
    """Device mesh express a view of topology of devices, represented by an n-d matrix of device ids.

    Parameters
    ----------
    shape: Union[ShapeTuple, List[int], Tuple[int]]
        Logical shape of device mesh
    device_start: int
    device_end: int
    device_step: int
        range(device_start, device_end, device_step) represents the device id in the mesh
    """

    def __init__(self, shape: Union[ShapeTuple, List[int], Tuple[int]], device_start: int, device_end: int, device_step: int = 1):
        if isinstance(shape, (list, tuple)):
            shape = ShapeTuple(shape)
        self.__init_handle_by_constructor__(
            ffi.DeviceMesh, shape, device_start, device_end, device_step  # type: ignore
        )


@tvm._ffi.register_object("relax.distributed.Placement")
class Placement(Object):
    """ Describes how data is distributed in each dimension of the device mesh

    Parameters
    ----------
    text_format: str
        The text format of placement.
    """
    def __init__(self, text_format: str):
        self.__init_handle_by_constructor__(
            ffi.Placement, text_format  # type: ignore
        )
    

File: src/relax/ir/distributed_tensor.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <tvm/relax/distributed_tensor.h>

namespace tvm {
namespace relax {

DeviceMesh::DeviceMesh(ShapeTuple shape, Array<Integer> device_ids) {
  int prod = 1;
  for (int i = 0; i < static_cast<int>(shape.size()); i++) {
    prod *= shape[i];
  }
  ObjectPtr<DeviceMeshNode> n = make_object<DeviceMeshNode>();
  CHECK(prod == device_ids.size())
      << "The number of device ids must match the product of the shape";
  n->shape = std::move(shape);
  n->device_ids = std::move(device_ids);
  data_ = std::move(n);
}

DeviceMesh::DeviceMesh(ShapeTuple shape, Integer device_start, Integer device_end,
                       Integer device_step) {
  CHECK(device_step->value > 0);
  ObjectPtr<DeviceMeshNode> n = make_object<DeviceMeshNode>();
  Array<Integer> device_ids;
  for (int i = device_start->value; i < device_end->value; i += device_step->value) {
    device_ids.push_back(i);
  }
  int prod = 1;
  for (int i = 0; i < static_cast<int>(shape.size()); i++) {
    prod *= shape[i];
  }
  CHECK(prod == device_ids.size())
      << "The number of device ids must match the product of the shape";
  n->device_ids = std::move(device_ids);
  n->shape = std::move(shape);
  n->device_start = std::move(device_start);
  n->device_end = std::move(device_end);
  n->device_step = std::move(device_step);
  data_ = std::move(n);
}

TVM_REGISTER_NODE_TYPE(DeviceMeshNode);
TVM_REGISTER_GLOBAL("relax.distributed.DeviceMesh")
    .set_body_typed([](ShapeTuple shape, Integer start, Integer end, Integer step) {
      return DeviceMesh(shape, start, end, step);
    });

Placement::Placement(Array<Integer> dim_placement) {
  ObjectPtr<PlacementNode> n = make_object<PlacementNode>();
  n->dim_placement = std::move(dim_placement);
  data_ = std::move(n);
}

Placement::Placement(String text_format) {
  Array<Integer> dim_placement;
  std::stringstream ss(text_format);
  while (true) {
    char indicator = 0;
    ss >> indicator;
    if (ss.eof()) {
      break;
    }
    if (indicator == 'R') {
      dim_placement.push_back(-1);
    } else if (indicator == 'S') {
      char lbracket;
      ss >> lbracket;
      CHECK(lbracket == '[');
      std::string substr;
      getline(ss, substr, ']');
      std::stringstream ss2(substr);
      int dim;
      ss2 >> dim;
      dim_placement.push_back(dim);
      CHECK(ss2.eof()) << "Invalid placement format";
    } else {
      LOG(FATAL) << "Invalid placement format";
    }
  }
  ObjectPtr<PlacementNode> n = make_object<PlacementNode>();
  n->dim_placement = std::move(dim_placement);
  data_ = std::move(n);
}

TVM_REGISTER_NODE_TYPE(PlacementNode);
TVM_REGISTER_GLOBAL("relax.distributed.Placement").set_body_typed([](String text_format) {
  return Placement(text_format);
});
}  // namespace relax
}  // namespace tvm